### ISSUE TITLE
Different download bundle selection for bundle variants

### ISSUE THREAD

**Issue by DaTribe** [easydigitaldownloads/easy-digital-downloads#2612]
When creating a download that is a bundle it should be possible to change the downloads that are distributed with each "download bundle variant". This works hand in hand with issue: #2611.

**Comment by SDavisMedia** [comment]
Bump! Here's a good reason to get this going:

![screen shot 2016-08-22 at 9 15 52 am](https://cloud.githubusercontent.com/assets/2359131/17862196/12b6c24e-6849-11e6-9b8e-06bc6eeea18b.png)

It would be great if this stayed in 2.7. :+1:

**Comment by sunnyratilal** [comment]
Ready for some testing!

Merry Christmas! 🎅

**Comment by pippinsplugins** [comment]
@sunnyratilal I'm getting an undefined offset notice:
```
wp-content/plugins/easy-digital-downloads/includes/admin/downloads/metabox.php:561
```

**Comment by pippinsplugins** [comment]
Also the display is wrong.

Notice how the dropdown has the price options available but there is also a `Price Assignment` dropdown? That dropdown shouldn't be there.
![screen shot 2016-12-26 at 10 39 33 am](https://cloud.githubusercontent.com/assets/1034109/21484416/b2b74108-cb57-11e6-8210-7ee30797aad1.png)

**Comment by SDavisMedia** [comment]
@sunnyratilal you're a saint!

@pippinsplugins If the bundle itself is variable pricing (like the scenario [here](https://github.com/easydigitaldownloads/easy-digital-downloads/issues/2612#issuecomment-55485090)), wouldn't the Price Assignment column be used for that? That way you can tie child download price options to a price assignment of the bundle itself. Here's my config:

<img width="1032" alt="screenshot 2016-12-26 08 49 58" src="https://cloud.githubusercontent.com/assets/2359131/21484477/52a85338-cb48-11e6-8ae9-5e02499d022a.png">

Seems to be working well on my end:

<img width="760" alt="screenshot 2016-12-26 08 50 55" src="https://cloud.githubusercontent.com/assets/2359131/21484488/6f662892-cb48-11e6-82ae-1ae4899e4335.png">

**Comment by pippinsplugins** [comment]
Oh! I didn't think of it that way. @SDavisMedia you're right.

We do have an issue, though. The price assignment drop down shows on bundles that do not use variable prices.

**Comment by SDavisMedia** [comment]
@sunnyratilal I am also getting the aforementioned error if I don't set things up properly/in the right oder. Same issue as Pippin here: https://github.com/easydigitaldownloads/easy-digital-downloads/issues/2612#issuecomment-269225534

@pippinsplugins (and Sunny) I'm having issues with the price assignment dropdown too. Particularly _when_ it decides to reveal itself. It seems to me that the entire **Price Assignment** _column_ in the Download Files metabox should depend directly on the **Enable variable pricing** setting in the Download Price metabox.

**Comment by pippinsplugins** [comment]
>>>It seems to me that the entire Price Assignment column in the Download Files metabox should depend directly on the Enable variable pricing setting in the Download Price metabox... only revealing itself if the bundle is variable pricing.

I agree.

**Comment by sunnyratilal** [comment]
@pippinsplugins @SDavisMedia General for another test.

**Comment by SDavisMedia** [comment]
@sunnyratilal The UI is looking great. Can't find any visual/behavioral quirks at the moment.

I am having a problem using the `All` selection for a product added to the bundle, though. Here's an example configuration:

<img width="1026" alt="screenshot 2016-12-29 09 29 18" src="https://cloud.githubusercontent.com/assets/2359131/21549755/4a7e63b0-cda9-11e6-944f-f1f3c3fa6a9c.png">

No matter which bundle price option I purchase, I should also get "Basic Download," its files, and a license key for it. On the purchase receipt I get the license for it but nothing else:

<img width="738" alt="screenshot 2016-12-29 09 32 06" src="https://cloud.githubusercontent.com/assets/2359131/21549804/addbbb10-cda9-11e6-9978-124a28836836.png">

Purchase record does not include the Basic Download product as being purchased but it does show that the license was generated with that purchase:

<img width="1021" alt="screenshot 2016-12-29 09 33 06" src="https://cloud.githubusercontent.com/assets/2359131/21549832/ea6102de-cda9-11e6-9a83-eb7cb6447907.png">

-----

Payment meta if you need it:

```
_edd_payment_user_id => 1
_edd_payment_meta=> array(8) {
  ["key"]=>
  string(32) "f24ec97055eedbc363cab214554acf67"
  ["email"]=>
  string(20) "sdavis2702@gmail.com"
  ["date"]=>
  string(0) ""
  ["user_info"]=>
  array(5) {
    ["id"]=>
    int(1)
    ["email"]=>
    string(20) "sdavis2702@gmail.com"
    ["first_name"]=>
    string(4) "Sean"
    ["last_name"]=>
    string(5) "Davis"
    ["discount"]=>
    string(4) "none"
  }
  ["downloads"]=>
  array(1) {
    [0]=>
    array(3) {
      ["id"]=>
      int(1387)
      ["quantity"]=>
      int(1)
      ["options"]=>
      array(2) {
        ["quantity"]=>
        int(1)
        ["price_id"]=>
        string(1) "1"
      }
    }
  }
  ["cart_details"]=>
  array(1) {
    [0]=>
    array(10) {
      ["name"]=>
      string(21) "Issue 2612 Bundle One"
      ["id"]=>
      int(1387)
      ["item_number"]=>
      array(3) {
        ["id"]=>
        int(1387)
        ["quantity"]=>
        int(1)
        ["options"]=>
        array(2) {
          ["quantity"]=>
          int(1)
          ["price_id"]=>
          string(1) "1"
        }
      }
      ["item_price"]=>
      float(50)
      ["quantity"]=>
      int(1)
      ["discount"]=>
      float(0)
      ["subtotal"]=>
      float(50)
      ["tax"]=>
      float(0)
      ["fees"]=>
      array(0) {
      }
      ["price"]=>
      float(50)
    }
  }
  ["fees"]=>
  array(0) {
  }
  ["currency"]=>
  string(3) "USD"
}
_edd_payment_gateway => manual
_edd_payment_user_email => sdavis2702@gmail.com
_edd_payment_user_ip => ::1
_edd_payment_purchase_key => f24ec97055eedbc363cab214554acf67
_edd_payment_mode => live
_edd_payment_customer_id => 2
_edd_payment_total => 50
_edd_payment_tax => 0
_edd_completed_date => 2016-12-29 10:28:16
```

**Comment by sunnyratilal** [comment]
@SDavisMedia The purchase record is correct - only the bundle should be displayed.  It's a bug in the method that retrieves the files for the purchase. I thought I already addressed the issue but I'll investigate.

**Comment by sunnyratilal** [comment]
@SDavisMedia Fixed - ready for another test.

**Comment by SDavisMedia** [comment]
> The purchase record is correct - only the bundle should be displayed.

Oops! I swear I'm an idiot sometimes lol. I knew that.

All working on my end now. I really like the UI and as of now, I cannot break it. Big thumbs up from me. :+1:

**Comment by pippinsplugins** [comment]
`Price Assignment` is showing improperly for me on a bundle that does _not_ have variable prices enabled.
![screen shot 2017-01-02 at 7 13 01 pm](https://cloud.githubusercontent.com/assets/1034109/21598278/9d29949e-d11f-11e6-969a-5613e71d6086.png)

**Comment by pippinsplugins** [comment]
![screen shot 2017-01-02 at 7 13 45 pm](https://cloud.githubusercontent.com/assets/1034109/21598279/aa563ce4-d11f-11e6-811d-e7ae9f4248a9.png)

**Comment by pippinsplugins** [comment]
If I toggle variable pricing on and off, the `Price Assignment` column goes away as expected, so it's just on first load.

**Comment by sunnyratilal** [comment]
@pippinsplugins Fixed in 561cdd1ac36c9ec6cb0d215a063164683764e904

**Comment by sunnyratilal** [comment]
@pippinsplugins @SDavisMedia Could I get one more round of testing here please?

**Comment by SDavisMedia** [comment]
@sunnyratilal I left an inline comment on the latest commit.

I was not never able to replicate what @pippinsplugins experienced above, and it's still working properly for me after a few more tests.

So after the correction for the last commit, this is a thumbs up from me. :+1:

**Comment by pippinsplugins** [comment]
@cklosowski mind reviewing as well please?

**Comment by pippinsplugins** [comment]
UI is working properly for me now.

**Comment by cklosowski** [comment]
@pippinsplugins it's on my list. I'll get it now.

**Comment by cklosowski** [comment]
This isn't saving correcly on the first creation of the prices. When I initially created it (before hitting save) I created 3 variable prices, and added 3 different price IDs from a product to it, and the price assignments seemed to not save correcly on the first save...after clicking publish/save, all things saved correctly.

We've seen this behavior before, and it's usually due to something in the chosen library not resetting the indexes. I'll try and replicate it again, but might be worth a look to see if you can replicate in creating an entire product in a single page load.

**Comment by sunnyratilal** [comment]
@cklosowski Just replicated it - I'll see if I can debug it and fix.

**Comment by cklosowski** [comment]
Getting some errors here now:
```
( ! ) Notice: Undefined variable: price_key in /Users/cklosowski/Development/edd.dev/wp-content/plugins/easy-digital-downloads/includes/admin/downloads/metabox.php on line 604

( ! ) Notice: Undefined variable: price_key in /Users/cklosowski/Development/edd.dev/wp-content/plugins/easy-digital-downloads/includes/admin/downloads/metabox.php on line 608
```

**Comment by sunnyratilal** [comment]
@cklosowski Fixed in 6a37d33ff4ffe98aea063e79915c92cad560b37e

**Comment by cklosowski** [comment]
@sunnyratilal Something still not quite right here.

First setup, before hitting 'Publish/Save'
![screen shot 2017-01-25 at 12 07 54 am](https://cloud.githubusercontent.com/assets/1390124/22278565/68ab6254-e292-11e6-89a9-3c85fa379af9.png)

After hitting 'Update/Save':
![screen shot 2017-01-25 at 12 08 08 am](https://cloud.githubusercontent.com/assets/1390124/22278569/6ecda840-e292-11e6-9814-53e38ff15f04.png)

**Comment by sunnyratilal** [comment]
Replicated the bug - it's saving correctly initially. Verified in `wp_postmeta`. It's to do with the logic when displaying the Price Assignments.

**Comment by sunnyratilal** [comment]
@cklosowski should be fixed now

**Comment by cklosowski** [comment]
Backend UI is looking good. Front end seems to be bumping the delivered items by 1:

Backend:
![screen shot 2017-01-26 at 10 09 49 am](https://cloud.githubusercontent.com/assets/1390124/22336289/a72125b4-e3af-11e6-90aa-0610c3b0209c.png)

Frontend after purchase:
![screen shot 2017-01-26 at 10 09 55 am](https://cloud.githubusercontent.com/assets/1390124/22336293/ac443626-e3af-11e6-980c-cc820cad874c.png)

**Comment by sunnyratilal** [comment]
@SDavisMedia and @cklosowski have both tested and said everything works as expected.

**Comment by sunnyratilal** [comment]
Merged!

**Comment by sunnyratilal** [comment]
Merged!

**Comment by pippinsplugins** [comment]
This is triggering some undefined notices on purchase confirmation.

Undefined variable: bundle_item_id 

wp-content/plugins/easy-digital-downloads/templates/shortcode-receipt.php:184

Undefined variable: bundle_price_id

wp-content/plugins/easy-digital-downloads/templates/shortcode-receipt.php:184

That happened when purchasing a bundle and a non-bundle at the same time.

**Comment by sunnyratilal** [comment]
Fixed in d6d55317edb5454f2b147e894f8e5223d00d58c9
